lib/upipe-modules/upipe_block_to_sound.c

```diff
@@ -0,0 +1,336 @@
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/** @file
+ * @short Upipe block_to_sound module - converts incoming block urefs to outgoing sound urefs
+ *
+ */
+
+#include <upipe/ubase.h>
+#include <upipe/uprobe.h>
+#include <upipe/uref.h>
+#include <upipe/uref_dump.h>
+#include <upipe/upipe.h>
+#include <upipe/udict.h>
+#include <upipe/uref_block.h>
+#include <upipe/ubuf_block.h>
+#include <upipe/uref_block_flow.h>
+#include <upipe/ubuf_sound.h>
+#include <upipe/uref_sound_flow.h>
+#include <upipe/upipe_helper_upipe.h>
+#include <upipe/upipe_helper_urefcount.h>
+#include <upipe/upipe_helper_void.h>
+#include <upipe/upipe_helper_input.h>
+#include <upipe/upipe_helper_output.h>
+#include <upipe/upipe_helper_ubuf_mgr.h>
+#include <upipe-modules/upipe_block_to_sound.h>
+
+#include <stdlib.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <stdarg.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <math.h>
+#include <assert.h>
+
+/** @hidden */
+static bool upipe_block_to_sound_handle(struct upipe *upipe, struct uref *uref,
+                                struct upump **upump_p);
+
+/** upipe_upipe_block_to_sound structure */
+struct upipe_block_to_sound {
+    /** refcount management structure */
+    struct urefcount urefcount;
+
+    /** ubuf manager */
+    struct ubuf_mgr *ubuf_mgr;
+    /** flow format packet */
+    struct uref *flow_format;
+    /** ubuf manager request */
+    struct urequest ubuf_mgr_request;
+
+    /** temporary uref storage (used during urequest) */
+    struct uchain urefs;
+    /** nb urefs in storage */
+    unsigned int nb_urefs;
+    /** max urefs in storage */
+    unsigned int max_urefs;
+    /** list of blockers (used during urequest) */
+    struct uchain blockers;
+
+    /** output pipe */
+    struct upipe *output;
+    /** flow_definition packet */
+    struct uref *flow_def;
+    /** output state */
+    enum upipe_helper_output_state output_state;
+    /** list of output requests */
+    struct uchain request_list;
+
+    /** public upipe structure */
+    struct upipe upipe;
+
+    /**  data to set flow_def and ubuf */
+    uint8_t sample_size;
+    uint8_t sample_bits;
+    uint8_t planes;
+    uint8_t channels;
+
+};
+
+UPIPE_HELPER_UPIPE(upipe_block_to_sound, upipe, UPIPE_BLOCK_TO_SOUND_SIGNATURE);
+UPIPE_HELPER_UREFCOUNT(upipe_block_to_sound, urefcount, upipe_block_to_sound_free)
+UPIPE_HELPER_VOID(upipe_block_to_sound);
+UPIPE_HELPER_INPUT(upipe_block_to_sound, urefs, nb_urefs, max_urefs, blockers,
+                   upipe_block_to_sound_handle);
```

Comment:
Not sure you need to use this helper because your handle function always returns true. You may directly do your stuff in the input function.